File: protocol/src/ptt-semantic-state.test.ts
import { describe, expect, it } from "vitest";
import { pttSemanticState } from "./ptt-semantic-state";

const base = { listener: false, connected: true, recovering: false, remoteTalking: false, pttState: "IDLE" as const, error: false };

describe("PTT semantic state", () => {
  it("returns BUSY_REMOTE to READY after remote RX ends", () => {
    expect(pttSemanticState({ ...base, remoteTalking: true })).toBe("BUSY_REMOTE");
    expect(pttSemanticState(base)).toBe("READY");
  });

  it("keeps listener neutral and exposes preparing and errors", () => {
    expect(pttSemanticState({ ...base, listener: true })).toBe("RX_ONLY");
    expect(pttSemanticState({ ...base, pttState: "REQUESTING" })).toBe("PREPARING");
    expect(pttSemanticState({ ...base, error: true })).toBe("ERROR");
    expect(pttSemanticState({ ...base, connected: false })).toBe("OFFLINE");
    expect(pttSemanticState({ ...base, recovering: true })).toBe("RECOVERING");
    expect(pttSemanticState({ ...base, pttState: "TRANSMITTING" })).toBe("TALKING");
    expect(pttSemanticState({ ...base, pttState: "BUSY" })).toBe("BUSY_REMOTE");
  });
});

File: android/app/src/test/java/com/dennomuso/koeon/core/ptt/PttSemanticStateTest.kt
package com.dennomuso.koeon.core.ptt

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class PttSemanticStateTest {
    @Test fun `remote busy returns to ready after receive ends`() {
        assertEquals(PttSemanticState.BUSY_REMOTE, pttSemanticState(true, true, false, true, PttState.IDLE))
        assertEquals(PttSemanticState.READY, pttSemanticState(true, true, false, false, PttState.IDLE))
    }

    @Test fun `listener never appears ready and priority is stable`() {
        assertEquals(PttSemanticState.RX_ONLY, pttSemanticState(false, true, false, false, PttState.IDLE))
        assertEquals(PttSemanticState.ERROR, pttSemanticState(true, true, false, true, PttState.ERROR))
        assertEquals(PttSemanticState.RECOVERING, pttSemanticState(true, true, true, true, PttState.IDLE))
        assertEquals(PttSemanticState.OFFLINE, pttSemanticState(true, false, false, false, PttState.IDLE))
        assertEquals(PttSemanticState.TALKING, pttSemanticState(true, true, false, false, PttState.TRANSMITTING))
        assertEquals(PttSemanticState.PREPARING, pttSemanticState(true, true, false, false, PttState.REQUESTING_FLOOR))
        assertEquals(PttSemanticState.PREPARING, pttSemanticState(true, true, false, false, PttState.RELEASING))
        assertEquals(PttSemanticState.BUSY_REMOTE, pttSemanticState(true, true, false, false, PttState.BUSY))
    }

    @Test fun `remote speaker blocks touch headset and hardware eligibility`() {
        assertFalse(localPttEligible(true, true, true, true, remoteTalking = true))
        assertTrue(localPttEligible(true, true, true, true, remoteTalking = false))
    }
}
